### - 카카오 API : 키워드 검색

- 카카오 API 테스트
  - 앱 ID : 
  - 네이티브 앱 키 : 
  - REST API 키 : 
  - JavaScript 키 : 
  - Admin 키 : 
- 키워드 검색
  - 질의어를 통해 장소 검색 결과 반환
  - category_group_name, category_name, place_url

In [ ]:
import json
import requests
# REST 키
rest_api_key = ''
# 헤더
headers = {"Authorization" : "KakaoAK {}".format(rest_api_key)}
# 파라미터
params = {"query" : "제주도 맛집", "page" : 10}
url = "https://dapi.kakao.com/v2/local/search/keyword.json"

# GET을 이용하여 획득
res = requests.get(url, headers=headers, params=params)
# Json을 이용하여 해제
doc = json.loads(res.text)
# 200일 경우 정상
res.status_code

200

In [ ]:
# 결과(Dict형으로 표현)
# documets : 각 검색 결과 정보
#  - address_name(String) : 전체 지번 주소
#  - category_group_code(String) : 중요 카테고리만 그룹핑한 카테고리 그룹 코드
#  - category_group_name(String) : 중요 카테고리만 그룹핑한 카테고리 그룹명
#  - category_name(String) : 카테고리 이름
#  - distance(String) : 중심좌표까지의 거리 (단, x,y 파라미터를 준 경우에만 존재), 단위 meter
#  - id(String) : 장소 ID
#  - phone(String) : 전화번호
#  - place_name(String) : 장소명, 업체명
#  - place_url(String) : 장소 상세페이지 URL
#  - road_address_name(String) : 전체 도로명 주소
#  - x(String) : X 좌표값, 경위도인 경우 longitude (경도)
#  - y(String) : Y 좌표값, 경위도인 경우 latitude(위도)
# meta : 검색 결과 메타 정보
doc

{'documents': [{'address_name': '제주특별자치도 제주시 추자면 묵리',
   'category_group_code': 'AT4',
   'category_group_name': '관광명소',
   'category_name': '여행 > 관광,명소 > 섬',
   'distance': '',
   'id': '10839940',
   'phone': '064-728-4281',
   'place_name': '추자도',
   'place_url': 'http://place.map.kakao.com/10839940',
   'road_address_name': '',
   'x': '126.314847295503',
   'y': '33.9528630502803'},
  {'address_name': '제주특별자치도 제주시 추자면 대서리 4-9',
   'category_group_code': '',
   'category_group_name': '',
   'category_name': '금융,보험 > 금융서비스 > 협동,소비조합 > 수산업협동조합',
   'distance': '',
   'id': '12868902',
   'phone': '064-742-8193',
   'place_name': '추자도수협',
   'place_url': 'http://place.map.kakao.com/12868902',
   'road_address_name': '제주특별자치도 제주시 추자면 추자로 2',
   'x': '126.29819666571721',
   'y': '33.96255605970134'},
  {'address_name': '제주특별자치도 제주시 추자면 대서리 4-41',
   'category_group_code': '',
   'category_group_name': '',
   'category_name': '교통,수송 > 교통시설 > 항구,포구',
   'distance': '',
   'id': '8115207'

### - 네이버 API : 지역 검색

- 네이버 API 테스트
  - Client ID : 
  - Client Secret : 
- 지역 검색
  - 질의어를 통해 장소 검색 결과 반환
  - category, description 사용 가능

In [44]:
import json
import urllib
import requests
# REST 키
client_key = 'xzjv8j9BbPlDkjv3L4cP'
client_secret = 'qXSjpcpeln'
# 헤더
headers = {"X-Naver-Client-Id" : client_key, 
           "X-Naver-Client-Secret" : client_secret}
# 파라미터
encText = urllib.parse.quote("제주 소리섬박물관")
url = "https://openapi.naver.com/v1/search/local.json?query="+encText

# GET을 이용하여 획득
res = requests.get(url, headers=headers)
# Json을 이용하여 해제
doc = json.loads(res.text)
# 200일 경우 정상
res.status_code

200

In [45]:
# 결과(Dict형으로 표현)
# items : 각 검색 결과 정보
#  - title(String) : 업체, 기관명
#  - link(String) : 기관의 상세 정보가 제공되는 네이버 페이지의 하이퍼텍스트 link
#  - category(String) : 중요 카테고리만 그룹핑한 카테고리 그룹명
#  - description(String) : 기관명에 대한 설명을 제공
#  - telephone(String) : 하위 호환성을 위해 존재
#  - address(String) : 기관명의 주소를 제공
#  - roadAddress(String) : 기관명의 도로명 주소를 제공
#  - mapx(String) : x좌표를 제공한다. 제공값은 카텍좌표계 값으로 제공
#  - mapy(String) : y좌표를 제공한다. 제공값은 카텍 좌표계 값으로 제공
doc

{'lastBuildDate': 'Mon, 10 Jan 2022 10:55:09 +0900',
 'total': 0,
 'start': 1,
 'display': 0,
 'items': []}

### - 네이버 API : 이미지 검색

- 네이버 API 테스트
  - Client ID : 
  - Client Secret : 
- 지역 검색
  - 질의어를 통해 이미지 검색 결과 반환
  - image 링크 반환

In [14]:
import json
import urllib
import requests
# REST 키
client_key = 'xzjv8j9BbPlDkjv3L4cP'
client_secret = 'qXSjpcpeln'
# 헤더
headers = {"X-Naver-Client-Id" : client_key, 
           "X-Naver-Client-Secret" : client_secret}
# 파라미터
encText = urllib.parse.quote("성산일출봉")
url = "https://openapi.naver.com/v1/search/image?query="+encText

# GET을 이용하여 획득
res = requests.get(url, headers=headers)
# Json을 이용하여 해제
doc = json.loads(res.text)
# 200일 경우 정상
res.status_code

200

In [19]:
# 결과(Dict형으로 표현)
doc['items'][0]

{'title': '수백만 년 전, 자연이 선사한 선물오름 성산일출봉',
 'link': 'http://post.phinf.naver.net/MjAyMTAzMDRfMTI4/MDAxNjE0ODIyMzEyMjUy.UJoOkQGq43hZtMB3rhAs_V73lVh10vnEHGqMp-9xu-kg.akFfJf-nMnM1ZHGQk4g5TehGydfR4zvM9emo6L6S7asg.PNG/I4FcB_p87GevbFGBBn4O2DNGHAVo.jpg',
 'thumbnail': 'https://search.pstatic.net/common/?src=http://post.phinf.naver.net/MjAyMTAzMDRfMTI4/MDAxNjE0ODIyMzEyMjUy.UJoOkQGq43hZtMB3rhAs_V73lVh10vnEHGqMp-9xu-kg.akFfJf-nMnM1ZHGQk4g5TehGydfR4zvM9emo6L6S7asg.PNG/I4FcB_p87GevbFGBBn4O2DNGHAVo.jpg&type=b150',
 'sizeheight': '818',
 'sizewidth': '543'}

### - 한국관광공사 LOD

- Python에서 SPARQL을 사용하기 위해 `pip install sparqlwrapper`

- 기본 정보

#### 1. 한국관광공사 LOD 테스트

In [91]:
from SPARQLWrapper import SPARQLWrapper, JSON

def getInfo():
    sparql = SPARQLWrapper("http://data.visitkorea.or.kr/sparql")
    prefix = """
                PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
                PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
                PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
                PREFIX dc: <http://purl.org/dc/elements/1.1/>
                PREFIX owl: <http://www.w3.org/2002/07/owl#>
                PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
                PREFIX vi: <http://www.saltlux.com/transformer/views#>
                PREFIX kto: <http://data.visitkorea.or.kr/ontology/>
                PREFIX ktop: <http://data.visitkorea.or.kr/property/>
                PREFIX ids: <http://data.visitkorea.or.kr/resource/>
                PREFIX wgs: <http://www.w3.org/2003/01/geo/wgs84_pos#>
                PREFIX foaf: <http://xmlns.com/foaf/0.1/>
                PREFIX geo: <http://www.saltlux.com/geo/property#>
                PREFIX pf: <http://www.saltlux.com/DARQ/property#> 
             """

    select_state = """
                    SELECT * 
                    WHERE {
                        ?resource a kto:Place ;
                            rdfs:label ?name ;
                            ktop:category ?category . 
                        ?category skos:prefLabel ?category_name . 
                        FILTER langMatches( lang(?category_name), "KO" ) 
                    } limit 10
                   """
    sparql.setQuery(prefix + select_state)
    sparql.setReturnFormat(JSON)
    result = sparql.query().convert()

    for r in result['results']['bindings']:
        print(r['name']['value'], ':', r['category_name']['value'])
    # return result

In [92]:
getInfo()

응봉산(울진) : 산
서울특별시교육청 구로도서관 : 도서관
일민미술관 : 미술관/화랑
서대문자연사박물관 : 박물관
백덕산(평창) : 산
이중섭 미술관 : 기념관
양산시립박물관 : 박물관
야미도 : 섬
황매산(산청) : 산
전라북도교육문화회관 : 수련시설


#### 2. 한국관광공사 LOD : 이미지

- 이미지 정보 : 대표 이미지(가장 첫 이미지)만 반환

In [297]:
from SPARQLWrapper import SPARQLWrapper, JSON

def getImage(ids):
    sparql = SPARQLWrapper("http://data.visitkorea.or.kr/sparql")
    prefix = """
                PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
                PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
                PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
                PREFIX dc: <http://purl.org/dc/elements/1.1/>
                PREFIX owl: <http://www.w3.org/2002/07/owl#>
                PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
                PREFIX vi: <http://www.saltlux.com/transformer/views#>
                PREFIX kto: <http://data.visitkorea.or.kr/ontology/>
                PREFIX ktop: <http://data.visitkorea.or.kr/property/>
                PREFIX ids: <http://data.visitkorea.or.kr/resource/>
                PREFIX wgs: <http://www.w3.org/2003/01/geo/wgs84_pos#>
                PREFIX foaf: <http://xmlns.com/foaf/0.1/>
                PREFIX geo: <http://www.saltlux.com/geo/property#>
                PREFIX pf: <http://www.saltlux.com/DARQ/property#> 
             """

    select_state = """
                    SELECT *
                    WHERE {
                        ids:""" + str(ids) + """ a kto:Place ;
                            foaf:depiction ?depiction .
                        ?depiction foaf:depicts ?depicts ;
                            foaf:thumbnail ?thumbnail .
                    }
                   """
    sparql.setQuery(prefix + select_state)
    sparql.setReturnFormat(JSON)
    result = sparql.query().convert()
    # 가장 첫 이미지 주소 반환
    try :
        return result['results']['bindings'][0]['depiction']['value']
    except:
        return ''

In [298]:
getImage(125587)

'http://tong.visitkorea.or.kr/cms/resource/56/2012856_image2_1.jpg'

#### 3. 한국관광공사 LOD : 제주도 정보

- 제주도 정보 : 1362개
  - IDS
  - 이름
  - 카테고리
  - 개요
  - 타입
  - 주소

In [235]:
from SPARQLWrapper import SPARQLWrapper, JSON

def getJeju(ASC = True):
    sparql = SPARQLWrapper("http://data.visitkorea.or.kr/sparql")
    prefix = """
                PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
                PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
                PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
                PREFIX dc: <http://purl.org/dc/elements/1.1/>
                PREFIX owl: <http://www.w3.org/2002/07/owl#>
                PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
                PREFIX vi: <http://www.saltlux.com/transformer/views#>
                PREFIX kto: <http://data.visitkorea.or.kr/ontology/>
                PREFIX ktop: <http://data.visitkorea.or.kr/property/>
                PREFIX ids: <http://data.visitkorea.or.kr/resource/>
                PREFIX wgs: <http://www.w3.org/2003/01/geo/wgs84_pos#>
                PREFIX foaf: <http://xmlns.com/foaf/0.1/>
                PREFIX geo: <http://www.saltlux.com/geo/property#>
                PREFIX pf: <http://www.saltlux.com/DARQ/property#> 
             """

    # select_state = """
    #                 SELECT (COUNT(?name) AS ?CNT)
    #                 WHERE {
    #                     ?resource a kto:Place ;
    #                         rdfs:label ?name ;
    #                         ktop:address ?address .
    #                         FILTER (contains(?address, "제주") )
    #                 }
    #                """
    if ASC:
        select_state = """
                        SELECT *
                        WHERE {
                            ?resource a kto:Place ;
                                rdfs:label ?name ;
                                ktop:category ?category ;
                                dc:description ?description ;
                                ktop:address ?address .
                                FILTER (contains(?address, "제주"))
                            ?category rdfs:label ?category_name .
                            FILTER langMatches( lang(?category_name), "KO" )
                        }
                        ORDER BY ASC(?name)
                    """
    else:
        select_state = """
                        SELECT *
                        WHERE {
                            ?resource a kto:Place ;
                                rdfs:label ?name ;
                                ktop:category ?category ;
                                dc:description ?description ;
                                ktop:address ?address .
                                FILTER (contains(?address, "제주"))
                            ?category rdfs:label ?category_name .
                            FILTER langMatches( lang(?category_name), "KO" )
                        }
                        ORDER BY DESC(?name)
                    """
    sparql.setQuery(prefix + select_state)
    sparql.setReturnFormat(JSON)
    result = sparql.query().convert()

    # for r in result['results']['bindings']:
    #     print(r['name']['value'], ':', r['category_name']['value'])
    return result

- 테스트

In [233]:
test = getJeju()

- 내용 예시

'head': {'vars': ['resource', 'name', 'category', 'description', 'type', 'address', 'type_name', 'category_name']},  
'results': {'bindings': [{'resource': {'type': 'uri', 'value': 'http://data.visitkorea.or.kr/resource/130494'},  
'name': {'type': 'literal', 'xml:lang': 'ko', 'value': '이중섭 미술관'},  
'category': {'type': 'uri', 'value': 'http://data.visitkorea.or.kr/resource/A02060200'},  
'description': {'type': 'literal', 'xml:lang': 'ko', 'value': '※ 코로나바이러스감염증-19 공지사항 ※ 내용 : 사전예약제 (2020.06.18 ~ 별도안내시) → 코로나...'},  
'type': {'type': 'uri', 'value': 'http://data.visitkorea.or.kr/ontology/Attraction'},  
'address': {'type': 'literal', 'xml:lang': 'ko', 'value': '제주특별자치도 서귀포시 이중섭로 27-3'},  
'type_name': {'type': 'literal', 'xml:lang': 'ko', 'value': '명소'},  
'category_name': {'type': 'literal', 'xml:lang': 'ko', 'value': '기념관'}}  

- 첫 인덱스 값 확인

In [198]:
test0 = test['results']['bindings'][0]

In [199]:
test0['resource']['value'].split('/')[-1], test0['name']['value'], test0['category_name']['value'], test0['description']['value'][:50], test0['address']['value']

('2741519',
 '(주)제주글램핑',
 '야영장,오토캠핑장',
 '제주글램핑은 천혜의 자연 경관을 제주에서 즐길 수 있는 수목원 같은 숲속 글램핑장으로 국내',
 '제주특별자치도 제주시 구좌읍 중산간동로 1921')

#### 4. 한국관광공사 LOD : 제주도 정보 => DF화

- df 추출 함수 정의

In [305]:
import pandas as pd

def getJeju_df():
    ids = []
    name = []
    category_name = []
    description = []
    address = []
    img = []
    # 1000개 씩 결과를 가져오기 때문에 오름차순, 내림차순으로 정렬하여 확인
    jeju = getJeju(True)
    for value in jeju['results']['bindings']:
        ids.append(value['resource']['value'].split('/')[-1])
        name.append(value['name']['value'])
        category_name.append(value['category_name']['value'])
        description.append(value['description']['value'])
        address.append(value['address']['value'])
    last_ids = ids[-1]

    jeju = getJeju(False)
    for value in jeju['results']['bindings']:
        if value['resource']['value'].split('/')[-1] == last_ids:
            break
        ids.append(value['resource']['value'].split('/')[-1])
        name.append(value['name']['value'])
        category_name.append(value['category_name']['value'])
        description.append(value['description']['value'])
        address.append(value['address']['value'])
    
    for i in ids:
        img.append(getImage(i))

    jeju_df = pd.DataFrame({
        'ids' : ids,
        'name' : name,
        'category' : category_name,
        'description' : description,
        'address' : address,
        'img' : img,
    })

    return jeju_df

In [ ]:
test_df = getJeju_df()

- type 컬럼 추가

In [ ]:
from SPARQLWrapper import SPARQLWrapper, JSON
def getType(ids):
    type_list = []

    sparql = SPARQLWrapper("http://data.visitkorea.or.kr/sparql")
    prefix = """
                PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
                PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
                PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
                PREFIX dc: <http://purl.org/dc/elements/1.1/>
                PREFIX owl: <http://www.w3.org/2002/07/owl#>
                PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
                PREFIX vi: <http://www.saltlux.com/transformer/views#>
                PREFIX kto: <http://data.visitkorea.or.kr/ontology/>
                PREFIX ktop: <http://data.visitkorea.or.kr/property/>
                PREFIX ids: <http://data.visitkorea.or.kr/resource/>
                PREFIX wgs: <http://www.w3.org/2003/01/geo/wgs84_pos#>
                PREFIX foaf: <http://xmlns.com/foaf/0.1/>
                PREFIX geo: <http://www.saltlux.com/geo/property#>
                PREFIX pf: <http://www.saltlux.com/DARQ/property#> 
             """

    select_state = """
                    SELECT * 
                    WHERE {
                        ids:"""+str(ids)+""" a kto:Place ;
                            rdf:type ?type .
                        ?type rdfs:label ?type_name .
                        FILTER langMatches( lang(?type_name), "KO" ) 
                    }
                   """
    sparql.setQuery(prefix + select_state)
    sparql.setReturnFormat(JSON)
    result = sparql.query().convert()

    for r in result['results']['bindings']:
        type_list.append(r['type_name']['value'])
    
    return type_list

In [309]:
test_df['type'] = ''
for idx, row in test_df.iterrows():
    type_list = getType(row['ids'])

    test_df.loc[idx, 'type'] = ','.join(type_list)  

In [311]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1358 entries, 0 to 1357
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   ids          1358 non-null   object
 1   name         1358 non-null   object
 2   category     1358 non-null   object
 3   description  1358 non-null   object
 4   address      1358 non-null   object
 5   img          1358 non-null   object
 6   type         1358 non-null   object
dtypes: object(7)
memory usage: 74.4+ KB


- 주소가 제주도인 경우 모두 DataFrame로 추출 : 총 1358개
- 이미지가 없는 경우 459개, 있는 경우 899개

In [324]:
(test_df['img'] != '').value_counts()

True     899
False    459
Name: img, dtype: int64

- 데이터 저장

In [325]:
# test_df.to_excel('./data/220110/한국관광공사_LOD_제주도.xlsx')

### - 관광 / 숙박 분리

- 관광 / 숙박은 카테고리상 차이가 나기 때문에 분리해주도록 한다.

#### 1. 데이터 확인

- 필터링1 버전으로 실행
- 필터링2는 레저/스포츠 부분만 전처리를 진행하였기 때문에 관광/숙박 시설은 필터링1과 동일

In [ ]:
import pandas as pd

df = pd.read_excel('./data/제주도 장소(POI) - 필터링1.xlsx')
df.head()

위치좌표 X축값   위치좌표 Y축값      구분     장소명                       소재지
0  126.495979  33.486592  관광/숙박  코리아나호텔     제주특별자치도 제주시 연동 303-10
1  126.385364  33.470809  관광/숙박   수산유원지  제주특별자치도 제주시 애월읍 수산리 1918
2  126.771241  33.529109  관광/숙박     만장굴    제주특별자치도 제주시 구좌읍 월정리 41
3  126.941753  33.458386  관광/숙박   성산일출봉  제주특별자치도 서귀포시 성산읍 신천리 114
4  126.394439  33.408524  관광/숙박  제주경마공원  제주특별자치도 제주시 애월읍 소길리 1206

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7466 entries, 0 to 7465
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   위치좌표 X축값   7466 non-null   float64
 1   위치좌표 Y축값   7466 non-null   float64
 2   구분         7466 non-null   object 
 3   장소명        7466 non-null   object 
 4   소재지        7466 non-null   object 
dtypes: float64(2), object(3)
memory usage: 291.8+ KB


- 구분 컬럼으로 카운트 확인

In [ ]:
df['구분'].unique()

array(['관광/숙박', '문화/종교/예술', '공원/산/동.식물원', '레져/스포츠'], dtype=object)

In [ ]:
df['구분'].value_counts()

관광/숙박         4621
레져/스포츠        2230
문화/종교/예술       347
공원/산/동.식물원     268
Name: 구분, dtype: int64

#### 2. 관광 / 숙박 : 숙박 시설 확인

- 관광/숙박 컬럼 선택

In [ ]:
# all = df.query('구분 == "관광/숙박"')
all = df[df['구분'] == '관광/숙박']
all.head()

위치좌표 X축값   위치좌표 Y축값      구분     장소명                       소재지
0  126.495979  33.486592  관광/숙박  코리아나호텔     제주특별자치도 제주시 연동 303-10
1  126.385364  33.470809  관광/숙박   수산유원지  제주특별자치도 제주시 애월읍 수산리 1918
2  126.771241  33.529109  관광/숙박     만장굴    제주특별자치도 제주시 구좌읍 월정리 41
3  126.941753  33.458386  관광/숙박   성산일출봉  제주특별자치도 서귀포시 성산읍 신천리 114
4  126.394439  33.408524  관광/숙박  제주경마공원  제주특별자치도 제주시 애월읍 소길리 1206

In [ ]:
all.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4621 entries, 0 to 7465
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   위치좌표 X축값   4621 non-null   float64
 1   위치좌표 Y축값   4621 non-null   float64
 2   구분         4621 non-null   object 
 3   장소명        4621 non-null   object 
 4   소재지        4621 non-null   object 
dtypes: float64(2), object(3)
memory usage: 216.6+ KB


- 숙박 시설 확인
  - '호텔', '민박', '모텔', '여관', '리조트', '펜션', '여인숙', '별장', '팬션', '콘도', '호스텔', '하우스'
  - '--텔'로 끝나는 경우
  - 2480개

In [ ]:
accommodations = ['호텔', '민박', '모텔', '여관', '리조트', '펜션', '여인숙', '별장', '팬션', '콘도', '호스텔', '하우스']

accommodation_df = all.query(f'장소명.str.contains("{"|".join(accommodations)}") or 장소명.str.endswith("텔")', engine='python').copy()

- 숙박 시설 DataFrame

In [ ]:
accommodation_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2480 entries, 0 to 7465
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   위치좌표 X축값   2480 non-null   float64
 1   위치좌표 Y축값   2480 non-null   float64
 2   구분         2480 non-null   object 
 3   장소명        2480 non-null   object 
 4   소재지        2480 non-null   object 
dtypes: float64(2), object(3)
memory usage: 116.2+ KB


In [ ]:
len(accommodation_df['장소명'].unique())

2388

In [ ]:
# accommodation_df.to_excel('./accomodation.xlsx')

#### 3. 관광 / 숙박 : 관광 시설 확인

- 숙박 시설의 인덱스를 제거하여 관광 시설 판별

In [ ]:
tourism_df = all.drop(accommodation_df.index, axis=0).copy()
tourism_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2141 entries, 1 to 7454
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   위치좌표 X축값   2141 non-null   float64
 1   위치좌표 Y축값   2141 non-null   float64
 2   구분         2141 non-null   object 
 3   장소명        2141 non-null   object 
 4   소재지        2141 non-null   object 
dtypes: float64(2), object(3)
memory usage: 100.4+ KB


In [ ]:
len(tourism_df['장소명'].unique())

2067

In [ ]:
# tourism_df.to_excel('tourism.xlsx')

#### 4. 관광 / 숙박 : 관광 시설 추가 관찰

- 엑셀을 이용해 직접 확인

### - 중복 좌표 처리

#### 1. 데이터 확인

- 필터링2 버전으로 실행
- 같은 좌표가 2개 이상인 경우만 추가하여 확인

In [1]:
import pandas as pd

df = pd.read_excel('./data/제주도 장소(POI) - 필터링2.xlsx')
df.head()

위치좌표 X축값   위치좌표 Y축값      구분     장소명                       소재지
0  126.495979  33.486592  관광/숙박  코리아나호텔     제주특별자치도 제주시 연동 303-10
1  126.385364  33.470809  관광/숙박   수산유원지  제주특별자치도 제주시 애월읍 수산리 1918
2  126.771241  33.529109  관광/숙박     만장굴    제주특별자치도 제주시 구좌읍 월정리 41
3  126.941753  33.458386  관광/숙박   성산일출봉  제주특별자치도 서귀포시 성산읍 신천리 114
4  126.394439  33.408524  관광/숙박  제주경마공원  제주특별자치도 제주시 애월읍 소길리 1206

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6510 entries, 0 to 6509
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   위치좌표 X축값   6510 non-null   float64
 1   위치좌표 Y축값   6510 non-null   float64
 2   구분         6510 non-null   object 
 3   장소명        6510 non-null   object 
 4   소재지        6510 non-null   object 
dtypes: float64(2), object(3)
memory usage: 254.4+ KB


In [3]:
df.columns

Index(['위치좌표 X축값 ', '위치좌표 Y축값 ', '구분', '장소명', '소재지'], dtype='object')

In [4]:
include_index = []      # 탐색한 좌표
duplicate_index = []    # 중복 좌표

for idx, row in df.iterrows():
    # 중복 좌표가 아닌 경우 탐색
    if idx not in include_index:
        # 중복 좌표의 인덱스 확인
        temp = list(df[((df['위치좌표 X축값 '] == row['위치좌표 X축값 ']) & (df['위치좌표 Y축값 '] == row['위치좌표 Y축값 ']))].index)
        # 길이가 2 이상인 경우 : 중복 좌표가 존재함을 의미
        if len(temp) >=2 :
            include_index.extend(temp)
            duplicate_index.append(temp)
        # 길이가 1인 경우 : 중복 좌표가 존재하지 않음을 의미
        else:
            include_index.append(idx)

- 전체 인덱스 개수

In [5]:
len(include_index)

6510

- 중복 좌표 개수

In [6]:
len(duplicate_index)

1073

#### 2. 중복 좌표 삭제

In [7]:
total = 0
names = {}
for idx, index_list in enumerate(duplicate_index, start=1):
    total += len(index_list)    # 중복 좌표의 개수 확인
    names[index_list[0]] = []   # 중복 좌표의 다른 이름 추가
    for idx2, index in enumerate(index_list):
        if idx2 != 0:
            # 다른 이름 추가
            names[index_list[0]].append(df.loc[index, '장소명'])
            # 행 삭제
            df.drop(index, axis=0, inplace=True)

- 전체 중복 좌표 수

In [8]:
total

2634

- 삭제한 행의 수

In [9]:
total - len(duplicate_index)

1561

- 중복 컬럼 삭제 후 정보

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4949 entries, 0 to 6431
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   위치좌표 X축값   4949 non-null   float64
 1   위치좌표 Y축값   4949 non-null   float64
 2   구분         4949 non-null   object 
 3   장소명        4949 non-null   object 
 4   소재지        4949 non-null   object 
dtypes: float64(2), object(3)
memory usage: 232.0+ KB


In [11]:
df['구분'].value_counts()

관광/숙박         3701
레져/스포츠         805
문화/종교/예술       234
공원/산/동.식물원     209
Name: 구분, dtype: int64

In [12]:
# 추가 이름 컬럼 생성
# 중복 좌표의 '이름' 추가, 없을 경우 '.'으로 표현
df['Additional_Name'] = '.'
for idx, value in names.items():
    df.loc[idx, 'Additional_Name'] = ','.join(value)

In [13]:
df.head()

위치좌표 X축값   위치좌표 Y축값      구분     장소명                       소재지  \
0  126.495979  33.486592  관광/숙박  코리아나호텔     제주특별자치도 제주시 연동 303-10   
1  126.385364  33.470809  관광/숙박   수산유원지  제주특별자치도 제주시 애월읍 수산리 1918   
2  126.771241  33.529109  관광/숙박     만장굴    제주특별자치도 제주시 구좌읍 월정리 41   
3  126.941753  33.458386  관광/숙박   성산일출봉  제주특별자치도 서귀포시 성산읍 신천리 114   
4  126.394439  33.408524  관광/숙박  제주경마공원  제주특별자치도 제주시 애월읍 소길리 1206   

                  Additional_Name  
0                               .  
1                               .  
2  (거문오름용암동굴계)만장굴,(제주화산섬과용암동굴)만장굴  
3           일출봉,(제주화산섬과용암동굴)성산일출봉  
4                      경마공원,제주경마장

- 중복 좌표 삭제

In [ ]:
# df.to_excel('./data/제주도 장소(POI) - 필터링3.xlsx')

### - 데이터 분리(구분)

- 필터링3 활용
- 구분 컬럼별로 데이터를 분리하여 저장

In [13]:
import pandas as pd

df_poi = pd.read_excel("./data/제주도 장소(POI) - 필터링3.xlsx", index_col=0)
df_poi.head(2)

위치좌표 X축값   위치좌표 Y축값      구분     장소명                       소재지  \
0  126.495979  33.486592  관광/숙박  코리아나호텔     제주특별자치도 제주시 연동 303-10   
1  126.385364  33.470809  관광/숙박   수산유원지  제주특별자치도 제주시 애월읍 수산리 1918   

  Additional_Name  
0               .  
1               .

In [24]:
df_poi['구분'].unique()

array(['관광/숙박', '문화/종교/예술', '공원/산/동.식물원', '레져/스포츠'], dtype=object)

In [26]:
df_poi['구분'].unique()
for kind in df_poi['구분'].unique():
    df_poi[df_poi['구분'] == kind].to_excel('''./data/220110/POI_{}.xlsx'''.format(kind.replace('/', '_')))

### - 데이터 확인

#### 1. 각 엑셀 데이터 확인

- 장소 정보1

In [21]:
import pandas as pd

park_df = pd.read_excel('./data/220110/네이버 API 결과 추가/공원_장소처리.xlsx')
tour_df = pd.read_excel('./data/220110/네이버 API 결과 추가/관광지_장소처리.xlsx')
culture_df = pd.read_excel('./data/220110/네이버 API 결과 추가/문화예술_장소처리.xlsx')
accom_df = pd.read_excel('./data/220110/네이버 API 결과 추가/숙박_장소처리.xlsx')
leisure_df = pd.read_excel('./data/220110/네이버 API 결과 추가/레저_스포츠_장소처리.xlsx')

In [24]:
print(park_df.shape, tour_df.shape, culture_df.shape, accom_df.shape, leisure_df.shape)
print(sum([park_df.shape[0], tour_df.shape[0], culture_df.shape[0], accom_df.shape[0], leisure_df.shape[0]]))

(159, 9) (1208, 9) (160, 9) (1301, 9) (540, 9)
3368


#### 2. 각 POI 정보 결합

- 장소 정보2
  - 각 엑셀 데이터 결합

In [29]:
import pandas as pd

poi_df = pd.read_excel('./data/220111/POI_장소처리.xlsx')
poi_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3368 entries, 0 to 3367
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   위치좌표 X축값         3368 non-null   float64
 1   위치좌표 Y축값         3368 non-null   float64
 2   구분               3368 non-null   object 
 3   장소명              3368 non-null   object 
 4   소재지              3368 non-null   object 
 5   Additional_Name  3368 non-null   object 
dtypes: float64(2), object(4)
memory usage: 158.0+ KB


#### 3. 식당 데이터

- 식당 정보

In [27]:
restaurant_df = pd.read_excel('./data/식당/제주도 식당 정보_전체.xlsx')
restaurant_df.shape

(20096, 39)

In [28]:
restaurant_df.columns

Index(['상가업소번호', '상호명', '지점명', '상권업종대분류코드', '상권업종대분류명', '상권업종중분류코드',
       '상권업종중분류명', '상권업종소분류코드', '상권업종소분류명', '표준산업분류코드', '표준산업분류명', '시도코드',
       '시도명', '시군구코드', '시군구명', '행정동코드', '행정동명', '법정동코드', '법정동명', '지번코드',
       '대지구분코드', '대지구분명', '지번본번지', '지번부번지', '지번주소', '도로명코드', '도로명', '건물본번지',
       '건물부번지', '건물관리번호', '건물명', '도로명주소', '구우편번호', '신우편번호', '동정보', '층정보',
       '호정보', '경도', '위도'],
      dtype='object')

### - 레져/스포츠 필요없는 데이터 삭제

#### 1. 필요없는 장소 삭제

- ['헬스', '휘트니스', '리조트', '연수원', '당구', '탁구', '볼링', '수련', '체육', '교육원', '테니스', '대피소', '주차장', '매표소', '요가', '게이트볼', '유도', '학교', '수양', '훈련', '휴게소', '사무소', '악지소'] 데이터 삭제

In [30]:
import pandas as pd

poi_df = pd.read_excel('./data/220111/POI_장소처리.xlsx')

In [40]:
poi_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3368 entries, 0 to 3367
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   위치좌표 X축값         3368 non-null   float64
 1   위치좌표 Y축값         3368 non-null   float64
 2   구분               3368 non-null   object 
 3   장소명              3368 non-null   object 
 4   소재지              3368 non-null   object 
 5   Additional_Name  3368 non-null   object 
dtypes: float64(2), object(4)
memory usage: 158.0+ KB


In [31]:
poi_df['구분'].unique()

array(['공원/산/동.식물원', '관광지', '레져/스포츠', '문화/종교/예술', '숙박'], dtype=object)

In [39]:
leisure = ['헬스', '휘트니스', '리조트', '연수원', '당구', '탁구', '볼링', '수련', '체육', '교육원', '테니스', '대피소', '주차장', '매표소', '요가', '게이트볼', '유도', '학교', '수양', '훈련', '휴게소', '사무소', '악지소']
leisure_df = poi_df.query('구분 == "레져/스포츠"')
drop_index = leisure_df.query(f'장소명.str.contains("{"|".join(leisure)}")', engine='python').index
len(drop_index)

178

In [41]:
poi_df.drop(drop_index, axis=0, inplace=True)
poi_df.reset_index(drop=True, inplace=True)

In [42]:
poi_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3190 entries, 0 to 3189
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   위치좌표 X축값         3190 non-null   float64
 1   위치좌표 Y축값         3190 non-null   float64
 2   구분               3190 non-null   object 
 3   장소명              3190 non-null   object 
 4   소재지              3190 non-null   object 
 5   Additional_Name  3190 non-null   object 
dtypes: float64(2), object(4)
memory usage: 149.7+ KB


In [43]:
# poi_df.to_excel('./data/220111/POI_장소처리2.xlsx', index=False)

#### 2. 중복 장소 삭제

- 엑셀을 통해 직접 눈으로 보며 같은 장소이면서 좌표가 다른 경우 삭제

In [45]:
import pandas as pd

poi_df3 = pd.read_excel('./data/220111/POI_장소처리3.xlsx')

In [46]:
poi_df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3068 entries, 0 to 3067
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   위치좌표 X축값         3068 non-null   float64
 1   위치좌표 Y축값         3068 non-null   float64
 2   구분               3068 non-null   object 
 3   장소명              3068 non-null   object 
 4   소재지              3068 non-null   object 
 5   Additional_Name  3068 non-null   object 
dtypes: float64(2), object(4)
memory usage: 143.9+ KB


### - 필요없는 데이터 삭제

- 문화/종교/예술, 공원/산/동.식물원, 숙박, 관광지 데이터 중 필요없는 데이터 삭제
  - 같은 장소이지만 경도, 위도가 다른 경우
  - 관광지가 아닌 경우 (ex) 놀이터(어린이 공원), 동네 운동장 등

### - ㅇ